File: src/io/github/hulang1024/chinesechess/scene/chessplay/rule/chess/ChessR.java
package io.github.hulang1024.chinesechess.scene.chessplay.rule.chess;

import io.github.hulang1024.chinesechess.scene.chessplay.rule.ChessPosition;
import io.github.hulang1024.chinesechess.scene.chessplay.rule.HostEnum;
import io.github.hulang1024.chinesechess.scene.chessplay.rule.MoveRules;
import io.github.hulang1024.chinesechess.scene.chessplay.rule.RoundGame;

/**
 * 车
 * @author Hu Lang
 */
public class ChessR extends AbstractChess {
    public ChessR(ChessPosition pos, HostEnum host) {
        super(pos, host);
    }

    @Override
    public boolean canGoTo(ChessPosition destPos, RoundGame game) {
        int rowOffset = destPos.row - pos.row;
        int colOffset = destPos.col - pos.col;

        // 进退都可以，但必须直线走
        if (!MoveRules.isStraightLineMove(rowOffset, colOffset, MoveRules.MAX_DISTANCE)) {
            return false;
        }

        // 只要到目标位置之间没有棋子
        if (Math.abs(rowOffset) > 0) {
            for (int row = pos.row; row != destPos.row; row += rowOffset) {
                if (!game.chessboard.isEmpty(row, pos.col)) {
                    return false;
                }
            }
        } else {
            for (int col = pos.col; col != destPos.col; col += colOffset) {
                if (!game.chessboard.isEmpty(pos.row, col)) {
                    return false;
                }
            }
        }

        return true;
    }
}

File: client/src/rulesets/chinesechess/rule/ChessR.ts
import ChessPos from "./ChessPos";
import AbstractChess from "./AbstractChess";
import { MAX_DISTANCE, isStraightLineMove, sign } from "./move_rules";
import ChessboardState from "./ChessboardState";

/**
 * 车
 */
export default class ChessR extends AbstractChess {
  canGoTo(destPos: ChessPos, chessboardState: ChessboardState): boolean {
    const rowOffset = destPos.row - this.pos.row;
    const colOffset = destPos.col - this.pos.col;

    // 进退都可以，但必须直线走
    if (!isStraightLineMove(rowOffset, colOffset, MAX_DISTANCE)) {
      return false;
    }

    // 只要到目标位置之间没有棋子
    if (Math.abs(rowOffset) > 0) {
      const k = sign(rowOffset);
      for (let row = this.pos.row + k; row != destPos.row; row += k) {
        if (!chessboardState.isEmpty(row, this.pos.col)) {
          return false;
        }
      }
    } else {
      const k = sign(colOffset);
      for (let col = this.pos.col + k; col != destPos.col; col += k) {
        if (!chessboardState.isEmpty(this.pos.row, col)) {
          return false;
        }
      }
    }

    return true;
  }

  clone(): ChessR {
    const clone = new ChessR(this.pos, this.host);
    clone.front = this.front;
    return clone;
  }
}
